source/common/grpc/google_async_client_impl.cc

```diff
@@ -235,6 +235,13 @@ void GoogleAsyncStreamImpl::closeStream() {
 
 void GoogleAsyncStreamImpl::resetStream() {
   ENVOY_LOG(debug, "resetStream");
+  // The gRPC API requires calling Finish() at the end of a stream, even
+  // if the stream is cancelled.
+  if (!finish_pending_) {
+    finish_pending_ = true;
+    rw_->Finish(&status_, &finish_tag_);
```

Comment:
Not sure I fully follow the various paths running through `resetStream()`:
1. We might end up here during destruction, what happens if the CQ is already being drained?
2. `cleanup()` can issues a `TryCancel` after the `Finish`. Is this legit?

Reply:
> Not sure I fully follow the various paths running through `resetStream()`:

I'm not sure I do either. :)  I suspect you're a lot more familiar with this code than I am.

> 1. We might end up here during destruction, what happens if the CQ is already being drained?

How would we wind up here during destruction?  I don't see that `~GoogleAsyncStreamImpl()` calls `resetStream()`.  Or are you referring to the fact that `~GoogleAsyncClientImpl()` calls `resetStream()` on the streams it owns?  I'm guessing the latter.

Note that the CQ is actually owned by `GoogleAsyncClientThreadLocal`, but it's not clear to me what the relationship is between `GoogleAsyncClientImpl` and `GoogleAsyncClientThreadLocal`; the former has a reference to the latter but doesn't appear to own it, and I don't know if there can be more than one `GoogleAsyncClientImpl` using the same `GoogleAsyncClientThreadLocal`.  Is there a guarantee that `GoogleAsyncClientThreadLocal` will not be destroyed until after all of the `GoogleAsyncClientImpl` objects that use it have been destroyed?

If there is such a guarantee, then I think this code is fine, because the destruction of `GoogleAsyncClientImpl` will result in calling `Finish()` on all of the streams, and only after that would `GoogleAsyncClientThreadLocal` be destroyed, which is what will shut down the CQ.  But if there is no such guarantee, then there is a problem, because the CQ would be shut down before we know that any new events will be started on it -- but I'm guessing that that's not actually the case, because if it was, it would presumably already have been an issue with callers continuing to use active streams after the CQ was shut down.

It's also worth noting one possible point of confusion here: The `draining_cq_` data member of `GoogleAsyncStreamImpl` is actually unrelated to whether the CQ itself is currently being shut down inside of `GoogleAsyncClientThreadLocal`.  As far as I can tell, that data member really just indicates whether the call has been terminated (either via `resetStream()` or by receiving the completion for a `Finish()` op) and is just waiting for any remaining async op completions to be returned from the queue before destroying the `GoogleAsyncStreamImpl` object.

> 2. `cleanup()` can issues a `TryCancel` after the `Finish`. Is this legit?

Yes, that's not a problem.  It's fine to call `TryCancel()` at any point during the call's lifecycle.

Reply:
`GoogleAsyncClientThreadLocal` is owned by the TLS slot (on a per-thread basis) established by `AsyncClientManagerImpl`. It would be pretty broken for a `AsyncClientImpl` to outlive the `AsyncClientManagerImpl` because of this reference relationship, so I think it should be assumed that in general `AsyncClientImpl` is gone by time we destruct `GoogleAsyncClientThreadLocal` (but this is not actually structurally enforced). I think I agree this is safe.

Re: code familiarity, I wrote this a few years back, so I'm actually paging a fair bit back in as we go through this exercise :)